### PR TITLE
Contract V1 Mainnet (Gas Optimization)

### PR DESCRIPTION
**Pull request by fabrice-quadrata**
## Description
This release includes a version of the Quadrata contracts that are better optimized for bulk query of attributes.
Included in this release are:
- rename `calculatePaymentETH` -> `queryFee`
- rename `getAttributesETH` -> `getAttributes`
- Adding new function like `getAttributesBulk` and `getAttributesBulkLegacy`.
- Add support for multi-chain (including chainId)
- Add ability to set their own minting price
- remove `mintPassport` and instead use `setAttributes`
- add PAUSABLE features
- Remove ability to pay with ERC20
- Add attribute permissions for issuers
- Optimize the way we store attributes
- Add deployment scripts
- Add post-deployment Verification scripts
- Add Migration scripts (to migrate old V1 -> new V2 contracts)

## Checklist
1. [x] Unit Testing?
2. [x] Backward Compatible?
3. [x] Upgrade Safe (storage)?
4. [x] Updated Confluence?
5. [x] Updated Gitbook?

### CODE REVIEW

File: contracts/QuadGovernance.sol [hunk 1]
```diff
@@ -271,197 +199,171 @@ contract QuadGovernance is IQuadGovernance, AccessControlUpgradeable, UUPSUpgrad
...
+    function setIssuerStatus(address _issuer, bool _status) external {
         require(hasRole(GOVERNANCE_ROLE, _msgSender()), "INVALID_ADMIN");
         require(_issuer != address(0), "ISSUER_ADDRESS_ZERO");
 
-        Issuer memory oldIssuerData = _issuers[_issuerIndices[_issuer]-1];
-        _issuers[_issuerIndices[_issuer]-1] = Issuer(oldIssuerData.issuer, _status);
+        _issuerStatus[_issuer] = _status;
 
-        if(_status == IssuerStatus.ACTIVE) {
+        if(_status == true) {
```
Comment (teddy-quadrata) [inline contracts/QuadGovernance.sol:228]:
```suggestion
        if(_status) {
```

Comment (teddy-quadrata) [inline contracts/QuadGovernance.sol:228]: Haha if x == true ;)

File: contracts/QuadGovernance.sol [hunk 2]
```diff
@@ -271,197 +199,171 @@ contract QuadGovernance is IQuadGovernance, AccessControlUpgradeable, UUPSUpgrad
...
+    }
+
+    /// @dev Get the revenue split between protocol and _issuers
+    /// @return ratio of revenue distribution
+    function revSplitIssuer() override public view returns(uint256) {
+        return _revSplitIssuer;
     }
 
     /// @dev Get an issuer at a certain index
+    /// @param _index Array index
```
Comment (teddy-quadrata) [inline contracts/QuadGovernance.sol:356]:
```suggestion
    /// @param _index array index
```

File: contracts/QuadPassport.sol [hunk 3]
```diff
@@ -1,340 +1,363 @@
...
     /// @dev initializer (constructor)
     /// @param _governanceContract address of the IQuadGovernance contract
-    /// @param _uri URI of the Quadrata Passport
     function initialize(
-        address _governanceContract,
-        string memory _uri
+        address _governanceContract
     ) public initializer {
         require(_governanceContract != address(0), "GOVERNANCE_ADDRESS_ZERO");
-        __ERC1155_init(_uri);
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:37]: Why did this get deleted?

File: contracts/QuadPassport.sol [hunk 4]
```diff
@@ -343,64 +366,101 @@ contract QuadPassport is IQuadPassport, ERC1155Upgradeable, UUPSUpgradeable, Qua
...
-        require(IAccessControlUpgradeable(address(governance)).hasRole(READER_ROLE, _msgSender()), "INVALID_READER");
-        if (!IAccessControlUpgradeable(address(governance)).hasRole(ISSUER_ROLE, _issuer))
-           return Attribute({value: bytes32(0), epoch: 0});
-        return _attributes[_account][_attribute][_issuer];
+    /// @dev Admin function to set Metadata URI to associate with a tokenId
+    /// @param _tokenId Token Id
+    /// @param _uri URI pointing to IPFS
+    function setTokenURI(uint256 _tokenId, string memory _uri) external override {
+        require(_msgSender() == address(governance), "ONLY_GOVERNANCE_CONTRACT");
+        _setURI(_uri, _tokenId);
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:374]: Set in initializer?

File: contracts/QuadPassport.sol [hunk 5]
```diff
@@ -343,64 +366,101 @@ contract QuadPassport is IQuadPassport, ERC1155Upgradeable, UUPSUpgradeable, Qua
...
-        require(IAccessControlUpgradeable(address(governance)).hasRole(READER_ROLE, _msgSender()), "INVALID_READER");
-        if (!IAccessControlUpgradeable(address(governance)).hasRole(ISSUER_ROLE, _issuer))
-           return Attribute({value: bytes32(0), epoch: 0});
-        return _attributesByDID[_dID][_attribute][_issuer];
+    /// @dev Admin function to pause critical operations (emergency)
+    function pause() external {
+        require(
+            IAccessControlUpgradeable(address(governance)).hasRole(PAUSER_ROLE, _msgSender()),
+            "INVALID_PAUSER"
+        );
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:382]:
```suggestion
        assert(
            IAccessControlUpgradeable(address(governance)).hasRole(PAUSER_ROLE, _msgSender()));
```

File: contracts/QuadPassport.sol [hunk 6]
```diff
@@ -343,64 +366,101 @@ contract QuadPassport is IQuadPassport, ERC1155Upgradeable, UUPSUpgradeable, Qua
...
-        uint256 _amount
-    ) public override {
-        require(IAccessControlUpgradeable(address(governance)).hasRole(READER_ROLE, _msgSender()), "INVALID_READER");
-        _accountBalancesETH[_account] += _amount;
+    /// @dev Admin function to unpause critical operations (emergency)
+    function unpause() external {
+        require(
+            IAccessControlUpgradeable(address(governance)).hasRole(PAUSER_ROLE, _msgSender()),
+            "INVALID_PAUSER"
+        );
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:391]:
```suggestion
        assert(
            IAccessControlUpgradeable(address(governance)).hasRole(PAUSER_ROLE, _msgSender())
        );
```

File: contracts/QuadPassport.sol [hunk 7]
```diff
@@ -343,64 +366,101 @@ contract QuadPassport is IQuadPassport, ERC1155Upgradeable, UUPSUpgradeable, Qua
...
-        require(IAccessControlUpgradeable(address(governance)).hasRole(READER_ROLE, _msgSender()), "INVALID_READER");
-        _accountBalances[_token][_account] += _amount;
+    /// @dev Retrieve the pause status of the contract
+    function passportPaused() external view override returns(bool) {
+        return paused();
     }
 
     function _authorizeUpgrade(address) internal view override {
-        require(IAccessControlUpgradeable(address(governance)).hasRole(GOVERNANCE_ROLE, _msgSender()), "INVALID_ADMIN");
+        require(
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:401]: Why not save on code size for these simple one liners? If it reverts, the reason can only be coming from 1 line. I don't see a need for using the more expensive require statement

File: contracts/QuadReader.sol [hunk 8]
```diff
@@ -39,519 +33,300 @@ import "./storage/QuadGovernanceStore.sol";
...
             }
-            passport.increaseAccountBalance(_tokenPayment, governance.treasury(), amountProtocol);
         }
-    }
 
+        require(_to != address(0), "WITHDRAW_ADDRESS_ZERO");
+        require(isValid, "WITHDRAWAL_ADDRESS_INVALID");
+        require(_amount <= address(this).balance, "INSUFFICIENT_BALANCE");
+        (bool sent,) = _to.call{value: _amount}("");
+        require(sent, "FAILED_TO_TRANSFER_NATIVE_ETH");
```
Comment (teddy-quadrata) [inline contracts/QuadReader.sol:306]: How is this line even testable?

**Review by teddy-quadrata — CHANGES_REQUESTED** [review]
A few nits but looks good overall

File: contracts/QuadReader.sol (hunk 8 above)
Comment (fabrice-quadrata) [reply contracts/QuadReader.sol:306]: not sure. I didn't change the behavior here. What do you suggest?

File: contracts/QuadPassport.sol (hunk 7 above)
Comment (fabrice-quadrata) [reply contracts/QuadPassport.sol:401]: this is an internal function call within many layers of calls.

Not too worried about saving on a few kb of codes size, better be intentional.

File: contracts/QuadPassport.sol (hunk 3 above)
Comment (fabrice-quadrata) [reply contracts/QuadPassport.sol:37]: Because we're not inheriting from ERC1155 anymore.

File: contracts/QuadPassport.sol (hunk 5 above)
Comment (fabrice-quadrata) [reply contracts/QuadPassport.sol:382]: No reasons to switch those to asserts to save almost nothing on gas.

Better be intentional about error messages.

File: contracts/QuadPassport.sol (hunk 4 above)
Comment (fabrice-quadrata) [reply contracts/QuadPassport.sol:374]: Nop, we are not setting anything by default anymore.
This isn't a global uri. it's per tokenId.

File: contracts/QuadGovernance.sol (hunk 1 above)
Comment (fabrice-quadrata) [reply contracts/QuadGovernance.sol:228]: will change in the next audit.

File: contracts/QuadReader.sol (hunk 8 above)
Comment (teddy-quadrata) [reply contracts/QuadReader.sol:306]: If we did want to test this line, we could withdraw to a contract that doesn't have a fallback. That should raise the error.

**Review by teddy-quadrata — APPROVED** [review]

**Review by faze-quadrata — APPROVED** [review]

**Pull request merged**